// @devexpress/analytics-core/core/internal/_dotnetTypesMapper.js
/**
* DevExpress Analytics (core\internal\_dotnetTypesMapper.js)
* Version:  24.2.5
* Build date: Feb 17, 2025
* Copyright (c) 2012 - 2025 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
export var DotnetTypes;
(function (DotnetTypes) {
    DotnetTypes["Enum"] = "Enum";
    DotnetTypes["SystemChar"] = "System.Char";
    DotnetTypes["SystemString"] = "System.String";
    DotnetTypes["SystemGuid"] = "System.Guid";
    DotnetTypes["SystemSByte"] = "System.SByte";
    DotnetTypes["SystemDecimal"] = "System.Decimal";
    DotnetTypes["SystemInt64"] = "System.Int64";
    DotnetTypes["SystemInt32"] = "System.Int32";
    DotnetTypes["SystemInt16"] = "System.Int16";
    DotnetTypes["SystemSingle"] = "System.Single";
    DotnetTypes["SystemDouble"] = "System.Double";
    DotnetTypes["SystemByte"] = "System.Byte";
    DotnetTypes["SystemUInt16"] = "System.UInt16";
    DotnetTypes["SystemUInt32"] = "System.UInt32";
    DotnetTypes["SystemUInt64"] = "System.UInt64";
    DotnetTypes["SystemBoolean"] = "System.Boolean";
    DotnetTypes["SystemDateTime"] = "System.DateTime";
    DotnetTypes["SystemDateOnly"] = "System.DateOnly";
    DotnetTypes["SystemTimeOnly"] = "System.TimeOnly";
    DotnetTypes["SystemTimeSpan"] = "System.TimeSpan";
    DotnetTypes["SystemObject"] = "System.Object";
    DotnetTypes["SystemBiteArray"] = "System.Byte[]";
    DotnetTypes["DevExpressExpression"] = "DevExpress.DataAccess.Expression";
})(DotnetTypes || (DotnetTypes = {}));
