// @devexpress/analytics-core/query-builder/wizard/pages/sqlDataSourceWizard/_sqlDataSourceWrapper.js
/**
* DevExpress Analytics (query-builder\wizard\pages\sqlDataSourceWizard\_sqlDataSourceWrapper.js)
* Version:  24.2.5
* Build date: Feb 17, 2025
* Copyright (c) 2012 - 2025 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
import { SqlDataSource } from '../../../dataSource/sql/sqlDataSource';
import { ModelSerializer } from '../../../../serializer/serializer';
import { Disposable } from '../../../../serializer/disposable';
export class _SqlDataSourceWrapper extends Disposable {
    get sqlQuery() {
        return this.sqlDataSource.queries()[this._queryIndex];
    }
    set sqlQuery(val) {
        if (val)
            val.parent = this.sqlDataSource;
        this.sqlDataSource.queries()[this._queryIndex] = val;
    }
    saveCustomQueries() {
        const serializer = new ModelSerializer();
        return this.customQueries.length > 0 && this.customQueries.map(x => JSON.stringify(serializer.serialize(x)));
    }
    save() {
        return JSON.stringify(new ModelSerializer().serialize(this.sqlDataSource));
    }
    constructor(sqlDataSourceJSON, queryName, requestWrapper) {
        super();
        this.sqlDataSourceJSON = sqlDataSourceJSON;
        this.customQueries = [];
        this.sqlDataSource = new SqlDataSource(sqlDataSourceJSON ? JSON.parse(sqlDataSourceJSON) : {}, undefined, requestWrapper);
        this._disposables.push(this.sqlDataSource);
        if (queryName) {
            this.sqlDataSource.queries().some((value, index) => {
                if (value.name() === queryName) {
                    this._queryIndex = index;
                    return true;
                }
                return false;
            });
        }
        else {
            this._queryIndex = this.sqlDataSource.queries().length;
        }
    }
}
