// @devexpress/analytics-core/core/widgets/_requiredNullableEditor.js
/**
* DevExpress Analytics (core\widgets\_requiredNullableEditor.js)
* Version:  24.2.5
* Build date: Feb 17, 2025
* Copyright (c) 2012 - 2025 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
import { Editor } from '../../property-grid/widgets/editor';
import { requiredValidationRules } from '../../property-grid/widgets/internal/_internal';
import { extend } from '../../serializer/_utils';
import { ValueEditorHelper } from '../../widgets/internal/_valueEditorHelper';
import { koUtils } from '../utils/_koUtils';
export class RequiredNullableEditor extends Editor {
    _getEditorValidationRules() {
        return (super._getEditorValidationRules() || []).concat(this.nullable ? [] : requiredValidationRules);
    }
    get nullable() {
        return koUtils.unwrap(this.editorOptions?.showClearButton);
    }
}
export function createNumericEditor(dotNetTypeFullName, specifics) {
    class DynamicNumberEditor extends RequiredNullableEditor {
        getOptions(templateOptions) {
            const options = super.getOptions(templateOptions);
            return ValueEditorHelper.getNumberEditorOptions(dotNetTypeFullName, specifics, options, this.nullable);
        }
    }
    return {
        header: 'dx-number-editor',
        editorType: DynamicNumberEditor
    };
}
export function createDateEditor(dotNetTypeFullName) {
    class DynamicDateEditor extends RequiredNullableEditor {
        getOptions(templateOptions) {
            const extendedOptions = extend(true, {}, templateOptions, ValueEditorHelper.editors[dotNetTypeFullName]);
            return super.getOptions(extendedOptions);
        }
    }
    return {
        header: 'dx-date',
        editorType: DynamicDateEditor
    };
}
