// @devexpress/analytics-core/core/widgets/registerDesignerEditors.native.js
/**
* DevExpress Analytics (core\widgets\registerDesignerEditors.native.js)
* Version:  24.2.5
* Build date: Feb 17, 2025
* Copyright (c) 2012 - 2025 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
import { editorTemplates } from '../../property-grid/widgets/editorsInfo';
import { GuidEditor } from '../../property-grid/widgets/guideditor';
import { DotnetTypes } from '../internal/_dotnetTypesMapper';
import { createDateEditor, createNumericEditor } from './_requiredNullableEditor';
import { ColorPickerEditor } from './colorPickerEditor';
let isRegistered = false;
export function registerNativeDesignerEditors() {
    if (!isRegistered) {
        isRegistered = true;
        editorTemplates.registerEditors({
            guid: { header: 'dxrd-guid', editorType: GuidEditor },
            date: createDateEditor(DotnetTypes.SystemDateTime),
            dateOnly: createDateEditor(DotnetTypes.SystemDateOnly),
            timeOnly: createDateEditor(DotnetTypes.SystemTimeOnly),
            nullableDate: { header: 'dx-date' },
            customColorEditor: { header: 'dxrd-colorpicker', editorType: ColorPickerEditor },
            sbyte: createNumericEditor(DotnetTypes.SystemSByte, 'integer'),
            decimal: createNumericEditor(DotnetTypes.SystemDecimal, 'float'),
            int64: createNumericEditor(DotnetTypes.SystemInt64, 'integer'),
            int32: createNumericEditor(DotnetTypes.SystemInt32, 'integer'),
            int16: createNumericEditor(DotnetTypes.SystemInt16, 'integer'),
            single: createNumericEditor(DotnetTypes.SystemSingle, 'float'),
            double: createNumericEditor(DotnetTypes.SystemDouble, 'float'),
            byte: createNumericEditor(DotnetTypes.SystemByte, 'integer'),
            uint16: createNumericEditor(DotnetTypes.SystemUInt16, 'integer'),
            uint32: createNumericEditor(DotnetTypes.SystemUInt32, 'integer'),
            uint64: createNumericEditor(DotnetTypes.SystemUInt64, 'integer')
        });
    }
}
