// devexpress-reporting/chart/internal/_requests.js
/**
* DevExpress HTML/JS Reporting (chart\internal\_requests.js)
* Version:  24.2.5
* Build date: Feb 17, 2025
* Copyright (c) 2012 - 2025 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
import { sendRequest } from '@devexpress/analytics-core/analytics-internal';
import { PrepareRequestArgs } from '@devexpress/analytics-core/analytics-utils';
import { HandlerUri } from '../_handlerUri';
export class ChartRequests {
    static getChartImage(uri, chartLayout, width, height) {
        return sendRequest(uri, 'chart', PrepareRequestArgs({
            width: width,
            height: height,
            Chart: JSON.stringify({
                'ChartXmlSerializer': {
                    '@version': '16.2.0.0',
                    Chart: chartLayout
                }
            })
        }));
    }
    static fieldListCallback(request) {
        const requestJson = PrepareRequestArgs(request);
        return sendRequest(HandlerUri(), 'fieldList', requestJson);
    }
}
