// @devexpress/analytics-core/query-builder/utils/requestwrapper.js
/**
* DevExpress Analytics (query-builder\utils\requestwrapper.js)
* Version:  24.2.5
* Build date: Feb 17, 2025
* Copyright (c) 2012 - 2025 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
import { sendRequest } from '../../core/utils/_utils.ajax';
import { HandlerUri } from '../initializer';
import { ModelSerializer } from '../../serializer/serializer';
import { serializeDataConnection } from './_dataConnection';
import { PrepareRequestArgs } from '../../core/internal/requestManager';
export class RequestWrapper {
    sendRequest(action, arg) {
        return sendRequest(HandlerUri(), action, arg);
    }
    _sendRequest(settings) {
        return sendRequest(settings);
    }
    getDbSchema({ connection, tables, getViews, getTables }) {
        const requestModel = {
            connectionJSON: serializeDataConnection(connection),
            tables: null,
            views: null,
            getTables,
            getViews
        };
        if (tables && tables.length > 0) {
            requestModel.tables = (tables || []).filter(x => !x.isView).map(x => x.name);
            requestModel.views = (tables || []).filter(x => x.isView).map(x => x.name);
        }
        return this.sendRequest('getDBSchema', PrepareRequestArgs(requestModel));
    }
    getDbStoredProcedures(connection) {
        const requestData = PrepareRequestArgs({
            connectionJSON: serializeDataConnection(connection)
        });
        return this.sendRequest('getDBStoredProcedures', requestData);
    }
    getSelectStatement(connection, queryJSON) {
        const requestData = PrepareRequestArgs({
            connectionJSON: serializeDataConnection(connection),
            sqlQueryJSON: queryJSON
        });
        return this.sendRequest('getSelectStatement', requestData);
    }
    getDataPreview(connection, queryJSON) {
        const requestData = PrepareRequestArgs({
            connectionJSON: serializeDataConnection(connection),
            sqlQueryJSON: queryJSON
        });
        return this.sendRequest('getDataPreview', requestData);
    }
    rebuildResultSchema(dataSource, queryName, relationsEditing = false, parameters, dataSerializationExtension) {
        const requestData = PrepareRequestArgs({
            sqlDataSourceJSON: JSON.stringify({ 'SqlDataSource': new ModelSerializer().serialize(dataSource) }),
            queryName: queryName,
            relationsEditing: relationsEditing,
            parameters: parameters,
            dataSerializationExtension: dataSerializationExtension
        });
        return this.sendRequest('rebuildResultSchema', requestData);
    }
    getFederationResultSchema(dataSource) {
        const serializedModel = dataSource.getSerializableModel().getSerializableFederationDataSourceInfo();
        const requestData = PrepareRequestArgs({
            federationDataSourceJSON: JSON.stringify({ 'FederationDataSource': new ModelSerializer().serialize(dataSource) }),
            dataSources: serializedModel.dataSources
        });
        return this.sendRequest('getFederationResultSchema', requestData);
    }
    validateJsonUri(jsonDataSource) {
        const uriJsonSourceJSON = JSON.stringify(jsonDataSource.source.serialize(true));
        const requestData = PrepareRequestArgs({
            uriJsonSourceJSON: uriJsonSourceJSON
        });
        const ajaxSettings = {
            uri: HandlerUri(),
            action: 'validateJsonEndPoint',
            arg: requestData,
            ignoreError: () => true
        };
        return this._sendRequest(ajaxSettings);
    }
    saveJsonSource(connectionName, jsonDataSource) {
        const jsonSource = jsonDataSource.source;
        const jsonSourceJSON = JSON.stringify(jsonSource.serialize(true));
        const requestData = PrepareRequestArgs({
            connectionName: connectionName,
            customJson: jsonSource.json(),
            uriJsonSourceJSON: jsonSourceJSON
        });
        return this.sendRequest('saveJsonSource', requestData);
    }
    getJsonSchema(jsonDataSource, parameters) {
        const jsonSource = jsonDataSource.source;
        const jsonSourceJSON = JSON.stringify(jsonSource.serialize(true));
        const requestData = PrepareRequestArgs({
            connectionName: jsonDataSource.connectionName(),
            customJson: jsonSource.json(),
            uriJsonSourceJSON: jsonSourceJSON,
            parameters: parameters
        });
        return this.sendRequest('getJsonSchema', requestData);
    }
    getObjectTypeDescriptions(context) {
        return this.sendRequest('getObjectSchema', context);
    }
}
