// @devexpress/analytics-core/widgets/criteria/operators/options/function.js
/**
* DevExpress Analytics (widgets\criteria\operators\options\function.js)
* Version:  24.2.5
* Build date: Feb 17, 2025
* Copyright (c) 2012 - 2025 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
export var FunctionOperatorType;
(function (FunctionOperatorType) {
    FunctionOperatorType[FunctionOperatorType["None"] = 0] = "None";
    FunctionOperatorType[FunctionOperatorType["Custom"] = 1] = "Custom";
    FunctionOperatorType[FunctionOperatorType["CustomNonDeterministic"] = 2] = "CustomNonDeterministic";
    FunctionOperatorType[FunctionOperatorType["Iif"] = 3] = "Iif";
    FunctionOperatorType[FunctionOperatorType["IsNull"] = 4] = "IsNull";
    FunctionOperatorType[FunctionOperatorType["IsNullOrEmpty"] = 5] = "IsNullOrEmpty";
    FunctionOperatorType[FunctionOperatorType["Trim"] = 6] = "Trim";
    FunctionOperatorType[FunctionOperatorType["Len"] = 7] = "Len";
    FunctionOperatorType[FunctionOperatorType["Substring"] = 8] = "Substring";
    FunctionOperatorType[FunctionOperatorType["Upper"] = 9] = "Upper";
    FunctionOperatorType[FunctionOperatorType["Lower"] = 10] = "Lower";
    FunctionOperatorType[FunctionOperatorType["Concat"] = 11] = "Concat";
    FunctionOperatorType[FunctionOperatorType["Ascii"] = 12] = "Ascii";
    FunctionOperatorType[FunctionOperatorType["Char"] = 13] = "Char";
    FunctionOperatorType[FunctionOperatorType["ToStr"] = 14] = "ToStr";
    FunctionOperatorType[FunctionOperatorType["Replace"] = 15] = "Replace";
    FunctionOperatorType[FunctionOperatorType["Reverse"] = 16] = "Reverse";
    FunctionOperatorType[FunctionOperatorType["Insert"] = 17] = "Insert";
    FunctionOperatorType[FunctionOperatorType["CharIndex"] = 18] = "CharIndex";
    FunctionOperatorType[FunctionOperatorType["Remove"] = 19] = "Remove";
    FunctionOperatorType[FunctionOperatorType["Abs"] = 20] = "Abs";
    FunctionOperatorType[FunctionOperatorType["Sqr"] = 21] = "Sqr";
    FunctionOperatorType[FunctionOperatorType["Cos"] = 22] = "Cos";
    FunctionOperatorType[FunctionOperatorType["Sin"] = 23] = "Sin";
    FunctionOperatorType[FunctionOperatorType["Atn"] = 24] = "Atn";
    FunctionOperatorType[FunctionOperatorType["Exp"] = 25] = "Exp";
    FunctionOperatorType[FunctionOperatorType["Log"] = 26] = "Log";
    FunctionOperatorType[FunctionOperatorType["Rnd"] = 27] = "Rnd";
    FunctionOperatorType[FunctionOperatorType["Tan"] = 28] = "Tan";
    FunctionOperatorType[FunctionOperatorType["Power"] = 29] = "Power";
    FunctionOperatorType[FunctionOperatorType["Sign"] = 30] = "Sign";
    FunctionOperatorType[FunctionOperatorType["Round"] = 31] = "Round";
    FunctionOperatorType[FunctionOperatorType["Ceiling"] = 32] = "Ceiling";
    FunctionOperatorType[FunctionOperatorType["Floor"] = 33] = "Floor";
    FunctionOperatorType[FunctionOperatorType["Max"] = 34] = "Max";
    FunctionOperatorType[FunctionOperatorType["Min"] = 35] = "Min";
    FunctionOperatorType[FunctionOperatorType["Acos"] = 36] = "Acos";
    FunctionOperatorType[FunctionOperatorType["Asin"] = 37] = "Asin";
    FunctionOperatorType[FunctionOperatorType["Atn2"] = 38] = "Atn2";
    FunctionOperatorType[FunctionOperatorType["BigMul"] = 39] = "BigMul";
    FunctionOperatorType[FunctionOperatorType["Cosh"] = 40] = "Cosh";
    FunctionOperatorType[FunctionOperatorType["Log10"] = 41] = "Log10";
    FunctionOperatorType[FunctionOperatorType["Sinh"] = 42] = "Sinh";
    FunctionOperatorType[FunctionOperatorType["Tanh"] = 43] = "Tanh";
    FunctionOperatorType[FunctionOperatorType["PadLeft"] = 44] = "PadLeft";
    FunctionOperatorType[FunctionOperatorType["PadRight"] = 45] = "PadRight";
    FunctionOperatorType[FunctionOperatorType["StartsWith"] = 46] = "StartsWith";
    FunctionOperatorType[FunctionOperatorType["EndsWith"] = 47] = "EndsWith";
    FunctionOperatorType[FunctionOperatorType["Contains"] = 48] = "Contains";
    FunctionOperatorType[FunctionOperatorType["ToInt"] = 49] = "ToInt";
    FunctionOperatorType[FunctionOperatorType["ToLong"] = 50] = "ToLong";
    FunctionOperatorType[FunctionOperatorType["ToFloat"] = 51] = "ToFloat";
    FunctionOperatorType[FunctionOperatorType["ToDouble"] = 52] = "ToDouble";
    FunctionOperatorType[FunctionOperatorType["ToDecimal"] = 53] = "ToDecimal";
    FunctionOperatorType[FunctionOperatorType["LocalDateTimeThisYear"] = 54] = "LocalDateTimeThisYear";
    FunctionOperatorType[FunctionOperatorType["LocalDateTimeThisMonth"] = 55] = "LocalDateTimeThisMonth";
    FunctionOperatorType[FunctionOperatorType["LocalDateTimeLastWeek"] = 56] = "LocalDateTimeLastWeek";
    FunctionOperatorType[FunctionOperatorType["LocalDateTimeThisWeek"] = 57] = "LocalDateTimeThisWeek";
    FunctionOperatorType[FunctionOperatorType["LocalDateTimeYesterday"] = 58] = "LocalDateTimeYesterday";
    FunctionOperatorType[FunctionOperatorType["LocalDateTimeToday"] = 59] = "LocalDateTimeToday";
    FunctionOperatorType[FunctionOperatorType["LocalDateTimeNow"] = 60] = "LocalDateTimeNow";
    FunctionOperatorType[FunctionOperatorType["LocalDateTimeTomorrow"] = 61] = "LocalDateTimeTomorrow";
    FunctionOperatorType[FunctionOperatorType["LocalDateTimeDayAfterTomorrow"] = 62] = "LocalDateTimeDayAfterTomorrow";
    FunctionOperatorType[FunctionOperatorType["LocalDateTimeNextWeek"] = 63] = "LocalDateTimeNextWeek";
    FunctionOperatorType[FunctionOperatorType["LocalDateTimeTwoWeeksAway"] = 64] = "LocalDateTimeTwoWeeksAway";
    FunctionOperatorType[FunctionOperatorType["LocalDateTimeNextMonth"] = 65] = "LocalDateTimeNextMonth";
    FunctionOperatorType[FunctionOperatorType["LocalDateTimeNextYear"] = 66] = "LocalDateTimeNextYear";
    FunctionOperatorType[FunctionOperatorType["LocalDateTimeTwoMonthsAway"] = 67] = "LocalDateTimeTwoMonthsAway";
    FunctionOperatorType[FunctionOperatorType["LocalDateTimeTwoYearsAway"] = 68] = "LocalDateTimeTwoYearsAway";
    FunctionOperatorType[FunctionOperatorType["LocalDateTimeLastMonth"] = 69] = "LocalDateTimeLastMonth";
    FunctionOperatorType[FunctionOperatorType["LocalDateTimeLastYear"] = 70] = "LocalDateTimeLastYear";
    FunctionOperatorType[FunctionOperatorType["LocalDateTimeYearBeforeToday"] = 71] = "LocalDateTimeYearBeforeToday";
    FunctionOperatorType[FunctionOperatorType["IsOutlookIntervalBeyondThisYear"] = 72] = "IsOutlookIntervalBeyondThisYear";
    FunctionOperatorType[FunctionOperatorType["IsOutlookIntervalLaterThisYear"] = 73] = "IsOutlookIntervalLaterThisYear";
    FunctionOperatorType[FunctionOperatorType["IsOutlookIntervalLaterThisMonth"] = 74] = "IsOutlookIntervalLaterThisMonth";
    FunctionOperatorType[FunctionOperatorType["IsOutlookIntervalNextWeek"] = 75] = "IsOutlookIntervalNextWeek";
    FunctionOperatorType[FunctionOperatorType["IsOutlookIntervalLaterThisWeek"] = 76] = "IsOutlookIntervalLaterThisWeek";
    FunctionOperatorType[FunctionOperatorType["IsOutlookIntervalTomorrow"] = 77] = "IsOutlookIntervalTomorrow";
    FunctionOperatorType[FunctionOperatorType["IsOutlookIntervalToday"] = 78] = "IsOutlookIntervalToday";
    FunctionOperatorType[FunctionOperatorType["IsOutlookIntervalYesterday"] = 79] = "IsOutlookIntervalYesterday";
    FunctionOperatorType[FunctionOperatorType["IsOutlookIntervalEarlierThisWeek"] = 80] = "IsOutlookIntervalEarlierThisWeek";
    FunctionOperatorType[FunctionOperatorType["IsOutlookIntervalLastWeek"] = 81] = "IsOutlookIntervalLastWeek";
    FunctionOperatorType[FunctionOperatorType["IsOutlookIntervalEarlierThisMonth"] = 82] = "IsOutlookIntervalEarlierThisMonth";
    FunctionOperatorType[FunctionOperatorType["IsOutlookIntervalEarlierThisYear"] = 83] = "IsOutlookIntervalEarlierThisYear";
    FunctionOperatorType[FunctionOperatorType["IsOutlookIntervalPriorThisYear"] = 84] = "IsOutlookIntervalPriorThisYear";
    FunctionOperatorType[FunctionOperatorType["IsThisWeek"] = 85] = "IsThisWeek";
    FunctionOperatorType[FunctionOperatorType["IsThisMonth"] = 86] = "IsThisMonth";
    FunctionOperatorType[FunctionOperatorType["IsThisYear"] = 87] = "IsThisYear";
    FunctionOperatorType[FunctionOperatorType["IsNextMonth"] = 88] = "IsNextMonth";
    FunctionOperatorType[FunctionOperatorType["IsNextYear"] = 89] = "IsNextYear";
    FunctionOperatorType[FunctionOperatorType["IsLastMonth"] = 90] = "IsLastMonth";
    FunctionOperatorType[FunctionOperatorType["IsLastYear"] = 91] = "IsLastYear";
    FunctionOperatorType[FunctionOperatorType["IsYearToDate"] = 92] = "IsYearToDate";
    FunctionOperatorType[FunctionOperatorType["IsSameDay"] = 93] = "IsSameDay";
    FunctionOperatorType[FunctionOperatorType["InRange"] = 94] = "InRange";
    FunctionOperatorType[FunctionOperatorType["InDateRange"] = 95] = "InDateRange";
    FunctionOperatorType[FunctionOperatorType["IsJanuary"] = 96] = "IsJanuary";
    FunctionOperatorType[FunctionOperatorType["IsFebruary"] = 97] = "IsFebruary";
    FunctionOperatorType[FunctionOperatorType["IsMarch"] = 98] = "IsMarch";
    FunctionOperatorType[FunctionOperatorType["IsApril"] = 99] = "IsApril";
    FunctionOperatorType[FunctionOperatorType["IsMay"] = 100] = "IsMay";
    FunctionOperatorType[FunctionOperatorType["IsJune"] = 101] = "IsJune";
    FunctionOperatorType[FunctionOperatorType["IsJuly"] = 102] = "IsJuly";
    FunctionOperatorType[FunctionOperatorType["IsAugust"] = 103] = "IsAugust";
    FunctionOperatorType[FunctionOperatorType["IsSeptember"] = 104] = "IsSeptember";
    FunctionOperatorType[FunctionOperatorType["IsOctober"] = 105] = "IsOctober";
    FunctionOperatorType[FunctionOperatorType["IsNovember"] = 106] = "IsNovember";
    FunctionOperatorType[FunctionOperatorType["IsDecember"] = 107] = "IsDecember";
    FunctionOperatorType[FunctionOperatorType["DateDiffTick"] = 108] = "DateDiffTick";
    FunctionOperatorType[FunctionOperatorType["DateDiffSecond"] = 109] = "DateDiffSecond";
    FunctionOperatorType[FunctionOperatorType["DateDiffMilliSecond"] = 110] = "DateDiffMilliSecond";
    FunctionOperatorType[FunctionOperatorType["DateDiffMinute"] = 111] = "DateDiffMinute";
    FunctionOperatorType[FunctionOperatorType["DateDiffHour"] = 112] = "DateDiffHour";
    FunctionOperatorType[FunctionOperatorType["DateDiffDay"] = 113] = "DateDiffDay";
    FunctionOperatorType[FunctionOperatorType["DateDiffMonth"] = 114] = "DateDiffMonth";
    FunctionOperatorType[FunctionOperatorType["DateDiffYear"] = 115] = "DateDiffYear";
    FunctionOperatorType[FunctionOperatorType["GetDate"] = 116] = "GetDate";
    FunctionOperatorType[FunctionOperatorType["GetMilliSecond"] = 117] = "GetMilliSecond";
    FunctionOperatorType[FunctionOperatorType["GetSecond"] = 118] = "GetSecond";
    FunctionOperatorType[FunctionOperatorType["GetMinute"] = 119] = "GetMinute";
    FunctionOperatorType[FunctionOperatorType["GetHour"] = 120] = "GetHour";
    FunctionOperatorType[FunctionOperatorType["GetDay"] = 121] = "GetDay";
    FunctionOperatorType[FunctionOperatorType["GetMonth"] = 122] = "GetMonth";
    FunctionOperatorType[FunctionOperatorType["GetYear"] = 123] = "GetYear";
    FunctionOperatorType[FunctionOperatorType["GetDayOfWeek"] = 124] = "GetDayOfWeek";
    FunctionOperatorType[FunctionOperatorType["GetDayOfYear"] = 125] = "GetDayOfYear";
    FunctionOperatorType[FunctionOperatorType["GetTimeOfDay"] = 126] = "GetTimeOfDay";
    FunctionOperatorType[FunctionOperatorType["Now"] = 127] = "Now";
    FunctionOperatorType[FunctionOperatorType["UtcNow"] = 128] = "UtcNow";
    FunctionOperatorType[FunctionOperatorType["Today"] = 129] = "Today";
    FunctionOperatorType[FunctionOperatorType["TruncateToMinute"] = 130] = "TruncateToMinute";
    FunctionOperatorType[FunctionOperatorType["IsSameHour"] = 131] = "IsSameHour";
    FunctionOperatorType[FunctionOperatorType["IsSameTime"] = 132] = "IsSameTime";
    FunctionOperatorType[FunctionOperatorType["BeforeMidday"] = 133] = "BeforeMidday";
    FunctionOperatorType[FunctionOperatorType["AfterMidday"] = 134] = "AfterMidday";
    FunctionOperatorType[FunctionOperatorType["IsNight"] = 135] = "IsNight";
    FunctionOperatorType[FunctionOperatorType["IsMorning"] = 136] = "IsMorning";
    FunctionOperatorType[FunctionOperatorType["IsAfternoon"] = 137] = "IsAfternoon";
    FunctionOperatorType[FunctionOperatorType["IsEvening"] = 138] = "IsEvening";
    FunctionOperatorType[FunctionOperatorType["IsLastHour"] = 139] = "IsLastHour";
    FunctionOperatorType[FunctionOperatorType["IsThisHour"] = 140] = "IsThisHour";
    FunctionOperatorType[FunctionOperatorType["IsNextHour"] = 141] = "IsNextHour";
    FunctionOperatorType[FunctionOperatorType["IsWorkTime"] = 142] = "IsWorkTime";
    FunctionOperatorType[FunctionOperatorType["IsFreeTime"] = 143] = "IsFreeTime";
    FunctionOperatorType[FunctionOperatorType["IsLunchTime"] = 144] = "IsLunchTime";
    FunctionOperatorType[FunctionOperatorType["AddTimeSpan"] = 145] = "AddTimeSpan";
    FunctionOperatorType[FunctionOperatorType["AddTicks"] = 146] = "AddTicks";
    FunctionOperatorType[FunctionOperatorType["AddMilliSeconds"] = 147] = "AddMilliSeconds";
    FunctionOperatorType[FunctionOperatorType["AddSeconds"] = 148] = "AddSeconds";
    FunctionOperatorType[FunctionOperatorType["AddMinutes"] = 149] = "AddMinutes";
    FunctionOperatorType[FunctionOperatorType["AddHours"] = 150] = "AddHours";
    FunctionOperatorType[FunctionOperatorType["AddDays"] = 151] = "AddDays";
    FunctionOperatorType[FunctionOperatorType["AddMonths"] = 152] = "AddMonths";
    FunctionOperatorType[FunctionOperatorType["AddYears"] = 153] = "AddYears";
    FunctionOperatorType[FunctionOperatorType["DateTimeFromParts"] = 154] = "DateTimeFromParts";
    FunctionOperatorType[FunctionOperatorType["DateOnlyFromParts"] = 155] = "DateOnlyFromParts";
    FunctionOperatorType[FunctionOperatorType["TimeOnlyFromParts"] = 156] = "TimeOnlyFromParts";
})(FunctionOperatorType || (FunctionOperatorType = {}));
